torch/_inductor/fx_passes/group_fusion.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import operator
+from typing import List
+
+import networkx
+import torch
+from torch._dynamo.utils import counters
+
+from ..pattern_matcher import (
+    CallFunctionVarArgs,
+    config_flag,
+    get_arg_value,
+    get_mutation_region_id,
+    stable_topological_sort,
+)
+
+log = logging.getLogger(__name__)
+
+
+def get_nx_graph_from_fx_graph(graph: torch.fx.Graph) -> networkx.DiGraph:
+    G = networkx.DiGraph()
+    for node in graph.nodes:
+        G.add_node(node)
+        for user in node.users:
+            G.add_edge(node, user)
+    return G
+
+
+class GroupFusionPass:
+    """
+    `GroupFusionPass` implements an Fx pass which groups together similar nodes into a fused op. Unlike `PatternMatcherPass`
+    which can only match local patterns in a graph, this pass can group together nodes across the graph. The only restriction
+    imposed is that the nodes being fused must not have a path in between each other (which would lead to a cycle). Additional
+    checks such as `extra_check` (at a node level), or `pair_check` can be added to pose additional restrictions on
+    nodes to be fused.
+    """
+
+    def __init__(
+        self,
+        pattern,
+        pair_check,
+        replacement_fn,
+        extra_check=lambda m: True,
+        prevent_match_across_mutations=False,
+    ):
+        """
+        Args:
+            pattern (torch.fx.Graph): A pattern to match nodes against. this pattern is used to filter nodes to be fused
+            pair_check (Callable[[torch.fx.Node, torch.fx.Node], bool]): A function to check if a pair of nodes can be fused
+            replacement_fn (Callable[[torch.fx.Graph, List[torch.fx.Node]]]): Function which does the actual replacement
+            extra_check (Callable[[torch.fx.Node], bool]): Extra check to filter the nodes
+            prevent_match_across_mutations (bool, optional): Prevent matching across mutations.
+        """
+        self.pattern = pattern
+        self.extra_check = extra_check
+        self.pair_check = pair_check
+        self.replacement_fn = replacement_fn
+        self.prevent_match_across_mutations = prevent_match_across_mutations
+
+    def apply(self, graph):
+        target_nodes = []
+        for node in graph.nodes:
+            if m := self.pattern.match(node):
+                if self.extra_check(m):
+                    target_nodes.append(node)
+
+        seen_nodes = set()
+        nx_graph = get_nx_graph_from_fx_graph(graph)
```

Comment:
```suggestion
        if len(target_nodes) <= 1:
            return
        nx_graph = get_nx_graph_from_fx_graph(graph)
```

To avoid compute the graph when it is not needed.